File: backend/app/api/admin/object_cache.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
# @author  : Liu Lijun
# @date    : 2026-05-29
# @description: Object Cache API Router 

from __future__ import annotations

from typing import Dict, Optional
from fastapi import APIRouter, Depends, Query, Request

from app.schemas.common import ApiResponse
from app.runtime.cache.registry import CacheRegistry
from app.core.security.auth_permission import AuthPermission

from app.schemas.admin.object_cache import (
    CacheInvalidateAllResponse,
    CacheInvalidateEverywhereRequest,
    CacheInvalidateEverywhereResponse,
    CacheInvalidateRequest,
    CacheInvalidateResponse,
    CacheStatsItem,
)

router = APIRouter()

def get_service(request: Request) -> CacheRegistry:
    return request.app.state.container.cache_registry

_list   = AuthPermission.Permission("object_cache:list")
_edit   = AuthPermission.Permission("object_cache:edit")

@router.get(
    "/list",
    response_model=ApiResponse[list[str]],
    summary="Get all registered cached names"
)
async def list_caches(
    svc:       CacheRegistry   = Depends(get_service),
    caller_id: int            = Depends(_list),   
) -> ApiResponse[list[str]]:
    data = svc.list_names()
    return ApiResponse(data=data)


@router.get(
    "/stats",
    response_model=ApiResponse[Dict[str, CacheStatsItem]],
    summary="Retrieve cache statistics; if name is empty, return all cached data.",   
)
async def get_cache_stats(
    name: Optional[str] = Query(None, description="Cache name; if empty, return statistics for all caches."),
    svc:       CacheRegistry   = Depends(get_service),
    caller_id: int            = Depends(_list), 
) -> ApiResponse[Dict[str, CacheStatsItem]]:

    stats = svc.stats(name)    
    return ApiResponse(data=stats)


@router.post(
    "/{name}/invalidate",
    response_model=ApiResponse[CacheInvalidateResponse],
    summary="Invalidate a specified record in the cache by key.",    
)
async def invalidate_cache_key(
    name: str,
    body: CacheInvalidateRequest,
    svc:       CacheRegistry   = Depends(get_service),
    caller_id: int            = Depends(_edit),
) -> ApiResponse[CacheInvalidateResponse]:
        
    invalidated = svc.invalidate(name, body.key)
    return ApiResponse(
        data=CacheInvalidateResponse(name=name, key=body.key, invalidated=invalidated)
    )

@router.post(
    "/{name}/invalidate-all",
    response_model=ApiResponse[CacheInvalidateAllResponse],
    summary="Invalidate all records in a specified cache.",    
)
async def invalidate_cache_all(
    name: str,
    svc:       CacheRegistry   = Depends(get_service),
    caller_id: int            = Depends(_edit),
) -> ApiResponse[CacheInvalidateAllResponse]:
  
    count = svc.invalidate_all(name)
    return ApiResponse(data=CacheInvalidateAllResponse(name=name, count=count))


@router.post(
    "/invalidate-everywhere",
    response_model=ApiResponse[CacheInvalidateEverywhereResponse],
    summary="Invalidate a specified record in all registered caches by key.",    
)
async def invalidate_cache_everywhere(
    body: CacheInvalidateEverywhereRequest,
    svc:       CacheRegistry   = Depends(get_service),
    caller_id: int            = Depends(_edit),
) -> ApiResponse[CacheInvalidateEverywhereResponse]:
    
    results = svc.invalidate_everywhere(body.key)
    return ApiResponse(
        data=CacheInvalidateEverywhereResponse(key=body.key, results=results)
    )

File: management/src/api/object_cache.ts
import { http } from "@/utils/http";
import { baseUrlApi } from "./utils";

/** 单个缓存的统计信息，对应后端 CacheStatsItem */
export interface CacheStatsItem {
  name: string;
  size: number;
  keys: any[];
  description: string;
}

export interface CacheInvalidateResult {
  name: string;
  key: any;
  invalidated: boolean;
}

export interface CacheInvalidateAllResult {
  name: string;
  count: number;
}

export interface CacheInvalidateEverywhereResult {
  key: any;
  results: Record<string, boolean>;
}

/** 获取所有已注册缓存的名称列表 */
export const getCacheNames = () => {
  return http.request<string[]>("get", baseUrlApi("admin/object-cache/list"));
};

/** 获取缓存统计信息；不传 name 时返回全部缓存 */
export const getCacheStats = (name?: string) => {
  return http.request<Record<string, CacheStatsItem>>(
    "get",
    baseUrlApi("admin/object-cache/stats"),
    { params: name ? { name } : {} }
  );
};

/** 按 key 失效指定缓存中的一条记录 */
export const invalidateCacheKey = (name: string, key: any) => {
  return http.request<CacheInvalidateResult>(
    "post",
    baseUrlApi(`admin/object-cache/${name}/invalidate`),
    { data: { key } }
  );
};

/** 清空指定缓存的全部记录 */
export const invalidateCacheAll = (name: string) => {
  return http.request<CacheInvalidateAllResult>(
    "post",
    baseUrlApi(`admin/object-cache/${name}/invalidate-all`)
  );
};

/** 使用同一个原始 key，尝试让所有已注册缓存失效 */
export const invalidateCacheEverywhere = (key: any) => {
  return http.request<CacheInvalidateEverywhereResult>(
    "post",
    baseUrlApi("admin/object-cache/invalidate-everywhere"),
    { data: { key } }
  );
};
